keras_nlp/models/gpt_neox/gpt_neox_decoder.py

```diff
@@ -0,0 +1,203 @@
+
+        if self._input_shape is not None:
+            self._build(self._input_shape)
+
+    def _build(self, input_shape):
+        # Create layers based on input shape.
+        self._built = True
+        self._input_shape = input_shape
+        # Infer the dimension of our hidden feature size from the build shape.
+        hidden_dim = input_shape[-1]
+
+        # Self attention layers.
+        self._self_attention_layer = GPTNeoXAttention(
+            num_heads=self.num_heads,
+            hidden_dim=hidden_dim,
+            max_position_embeddings=self.max_position_embeddings,
+            dropout=self.dropout,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            bias_initializer=clone_initializer(self.bias_initializer),
+            rotary_pct=self.rotary_pct,
+            rotary_emb_base=self.rotary_emb_base,
+        )
+
+        self._self_attention_layernorm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon,
+        )
+
+        self._self_attention_dropout = keras.layers.Dropout(
+            rate=self.dropout,
+        )
+
+        # Feedforward layers.
+        self._feedforward_intermediate_dense = keras.layers.Dense(
+            self.intermediate_dim,
+            activation=self.activation,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            bias_initializer=clone_initializer(self.bias_initializer),
+        )
+        self._feedforward_output_dense = keras.layers.Dense(
+            hidden_dim,
+            kernel_initializer=clone_initializer(self.kernel_initializer),
+            bias_initializer=clone_initializer(self.bias_initializer),
+        )
+        self._feedforward_layernorm = keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon,
+        )
+        self._feedforward_dropout = keras.layers.Dropout(
+            rate=self.dropout,
+        )
+
+    def call(
+        self,
+        decoder_sequence,
+        encoder_sequence=None,
+        decoder_padding_mask=None,
+        decoder_attention_mask=None,
+        encoder_padding_mask=None,
+        encoder_attention_mask=None,
+    ):
+
+        if not self._built:
+            self._build(decoder_sequence.shape)
+
+        #         has_self_attention_cache = self_attention_cache is not None
+
+        x = decoder_sequence  # Intermediate result.
+
+        # Compute self attention mask.
+        batch_size = tf.shape(decoder_sequence)[0]
+        input_length = output_length = tf.shape(decoder_sequence)[1]
+        # We need to handle a rectangular causal mask when doing cached
+        # decoding. For generative inference, `decoder_sequence` will
+        # generally be length 1, and `cache` will be the full generation length.
+        #         if self_attention_cache is not None:
+        #             input_length = tf.shape(self_attention_cache)[2]
+        self_attention_mask = compute_causal_mask(
```

Comment:
Hey @mattdangerw I'm not sure if I've correctly employed causal_mask in the model, could you take a look?

Outputs are not matching as of now. Also, I need to add caching mechanism in the model.

Reply:
@shivance - quick question: why would causal masking, caching, etc. be any different from GPT-2?

Reply:
> @shivance - quick question: why would causal masking, caching, etc. be any different from GPT-2?

@abheesht17 because call argument takes in `hidden_states` here which we project into query, key, value.

I'm slightly confused between relation of hidden_states and key (shouldn't `key = hidden_state[hidden_dim: 2* hidden_dim, :]` ? )

https://github.com/keras-team/keras-nlp/blob/3efdb8f5d3cd7a381fc4da3fdf1bd54c1b936760/keras_nlp/layers/cached_multi_head_attention.py#L93

Reply:
https://github.com/keras-team/keras-nlp/blob/3efdb8f5d3cd7a381fc4da3fdf1bd54c1b936760/keras_nlp/layers/transformer_decoder.py#L334-L340C10